r2/types.go

```diff
@@ -0,0 +1,308 @@
+// of its ruleset.
+type CurrentRuleSet struct {
+	Namespace    string
+	Version      int
+	CutoverNanos int64
+	MappingRules []*rules.MappingRuleView
+	RollupRules  []*rules.RollupRuleView
+}
+
+// nolint
+func newCurrentRuleSetView(r rules.RuleSet) (*CurrentRuleSet, error) {
+	mrs, err := r.MappingRules()
+	if err != nil {
+		return nil, err
+	}
+
+	rrs, err := r.RollupRules()
+	if err != nil {
+		return nil, err
+	}
+
+	currentMappingRules := newMappingRules(mrs)
+	currentRollupRules := newRollupRules(rrs)
+
+	return &CurrentRuleSet{
+		Namespace:    string(r.Namespace()),
+		Version:      r.Version(),
+		CutoverNanos: r.CutoverNanos(),
+		MappingRules: currentMappingRules,
+		RollupRules:  currentRollupRules,
+	}, nil
+}
+
+type ruleSetJSON struct {
+	Namespace     string            `json:"id"`
+	Version       int               `json:"version"`
+	CutoverMillis int64             `json:"cutoverMillis"`
+	MappingRules  []mappingRuleJSON `json:"mappingRules"`
+	RollupRules   []rollupRuleJSON  `json:"rollupRules"`
+}
+
+func (c CurrentRuleSet) ruleSetJSON() ruleSetJSON {
+	mrJSON := make([]mappingRuleJSON, len(c.MappingRules))
+	for i, m := range c.MappingRules {
+		mrJSON[i] = newMappingRuleJSON(m)
+	}
+	rrJSON := make([]rollupRuleJSON, len(c.RollupRules))
+	for i, r := range c.RollupRules {
+		rrJSON[i] = newRollupRuleJSON(r)
+	}
+	return ruleSetJSON{
+		Namespace:     c.Namespace,
+		Version:       c.Version,
+		CutoverMillis: nanoToMilli(c.CutoverNanos),
+		MappingRules:  mrJSON,
+		RollupRules:   rrJSON,
+	}
+}
+
+// nolint
+func newMappingRules(mrs rules.MappingRules) []*rules.MappingRuleView {
+	result := make([]*rules.MappingRuleView, len(mrs))
+	idx := 0
+	for _, m := range mrs {
+		if len(m) == 0 {
+			continue
+		}
+		result[idx] = m[len(m)-1]
+		idx++
+	}
+	return result
+}
+
+// nolint
+func newRollupRules(rrs rules.RollupRules) []*rules.RollupRuleView {
+	result := make([]*rules.RollupRuleView, len(rrs))
+	idx := 0
+	for _, r := range rrs {
+		if len(r) == 0 {
+			continue
+		}
+		result[idx] = r[len(r)-1]
+		idx++
+	}
+	return result
+}
+
+// API response
+type apiResponse struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+type namespaceJSON struct {
+	ID string `json:"id" validate:"required"`
+}
+
+// Namespace Level
+type namespacesJSON struct {
+	Version    int             `json:"version"`
+	Namespaces []namespaceJSON `json:"namespaces"`
+}
+
+type filters map[string]string
+
+func (f filters) MarshalJSON() ([]byte, error) {
+	tmp := make([]string, 0, len(f))
+	for k, v := range f {
+		tmp = append(tmp, fmt.Sprintf("%s:%s", k, v))
+	}
+	marshalled := strconv.Quote(strings.Join(tmp, " "))
+	return []byte(marshalled), nil
+}
+
+func (f *filters) UnmarshalJSON(data []byte) error {
+	str := string(data)
+	unquoted, err := strconv.Unquote(str)
+	if err != nil {
+		return err
+	}
+
+	*f = make(map[string]string)
```

Comment:
nit: I'd create a temp map here, and assign *f to it before the final return in case things error out and f end up in a weird state.